workflow/controller/steps.go

```diff
@@ -68,7 +68,7 @@ func (woc *wfOperationCtx) executeSteps(ctx context.Context, nodeName string, tm
 		{
 			sgNode, err := woc.wf.GetNodeByName(sgNodeName)
 			if err != nil {
-				_ = woc.initializeNode(sgNodeName, wfv1.NodeTypeStepGroup, stepTemplateScope, &wfv1.WorkflowStep{}, stepsCtx.boundaryID, wfv1.NodeRunning)
```

Comment:
Only top-level dag or step nodes should be marked as `hooked`.
I confirmed this case [See PR description]

Reply:
Sorry, can you clarify what you mean? I figure `hooked` means that the Node is a hook, right?

Are you referring here to the case in which the hook is itself a `StepGroup`? Is that your picture titled "When running hook node is Step"?

Reply:
Sorry for the little description. And thank you for pointing out that.
It's only a definition of `hooked` status.
As far as I understood, `StepGroup`/`DagTask` is children of `Step`/`Dag` node, and we cannot hook `StepGroup`/`DagTask` itself.

I mean the following situation:
![image](https://github.com/argoproj/argo-workflows/assets/83329336/8d212e1b-9a04-4673-985b-069b6fb3b898)
Only top-level Step nodes are marked as `Hooked: true` and children are not.

<details>
<summary>manifest</summary>

```yaml
spec:
  templates:
    - name: argosay
      inputs: {}
      outputs: {}
      metadata: {}
      container:
        name: ''
        image: argoproj/argosay:v2
        command:
          - /bin/sh
          - '-c'
        args:
          - /bin/sleep 2; exit 100; /argosay
        resources: {}
      retryStrategy:
        limit: 1
    - name: hook
      inputs: {}
      outputs: {}
      metadata: {}
      steps:
        - - name: A
            template: argosay2
            arguments: {}
        - - name: B
            template: argosay2
            arguments: {}
    - name: argosay2
      inputs: {}
      outputs: {}
      metadata: {}
      container:
        name: ''
        image: argoproj/argosay:v2
        command:
          - /bin/sh
          - '-c'
        args:
          - exit 0
        resources: {}
  entrypoint: argosay
  arguments: {}
  hooks:
    failed:
      template: hook
      arguments: {}
      expression: workflow.status == "Failed"
    running:
      template: hook
      arguments: {}
      expression: workflow.status == "Running"
    succeed:
      template: hook
      arguments: {}
      expression: workflow.status == "Succeeded"
```

</details>

This 'definition' might be useful if we would like to retry hooked nodes.
![image](https://github.com/argoproj/argo-workflows/assets/83329336/344a6ee0-c8e8-4dc2-9b93-a95085ee5894)

<details>
<summary>manifest</summary>

```
spec:
  templates:
    - name: argosay
      inputs: {}
      outputs: {}
      metadata: {}
      container:
        name: ''
        image: argoproj/argosay:v2
        command:
          - /bin/sh
          - '-c'
        args:
          - /bin/sleep 2; exit 100; /argosay
        resources: {}
      retryStrategy:
        limit: 1
    - name: hook
      inputs: {}
      outputs: {}
      metadata: {}
      steps:
        - - name: A
            template: argosay
            arguments: {}
        - - name: B
            template: argosay
            arguments: {}
    - name: argosay2
      inputs: {}
      outputs: {}
      metadata: {}
      container:
        name: ''
        image: argoproj/argosay:v2
        command:
          - /bin/sh
          - '-c'
        args:
          - exit 0
        resources: {}
      retryStrategy:
        limit: 1
  entrypoint: argosay
  arguments: {}
  hooks:
    failed:
      template: hook
      arguments: {}
      expression: workflow.status == "Failed"
    running:
      template: hook
      arguments: {}
      expression: workflow.status == "Running"
    succeed:
      template: hook
      arguments: {}
      expression: workflow.status == "Succeeded"
```

</details>

Reply:
thanks for drawing - one correction - I think "Type: Step" is supposed to be "Type: Steps", right? (I just tried running your example)

Reply:
Oh, yes exactly. Fixed.

Reply:
I looked carefully at your two examples above: I see - in the first one your hook calls a template which defines some steps. In your second example, your [template-level] hook calls a template which is a Pod.